src/napi-inl.h

```diff
@@ -0,0 +1,1484 @@
+﻿#ifndef SRC_NAPI_INL_H_
+#define SRC_NAPI_INL_H_
+
+////////////////////////////////////////////////////////////////////////////////
+// NAPI C++ Wrapper Classes
+//
+// Inline header-only implementations for "NAPI" ABI-stable C APIs for Node.js.
+////////////////////////////////////////////////////////////////////////////////
+
+#include <cassert>
+
+namespace Napi {
+
+// Adapt the NODE_MODULE registration function to NAPI:
+//  - Wrap the arguments in NAPI wrappers.
+//  - Catch any NAPI errors that might be thrown.
+#define NAPI_MODULE(modname, regfunc)                                                 \
+  void __napi_ ## regfunc(napi_env env, napi_value exports, napi_value module) {      \
+    try {                                                                             \
+      regfunc(Napi::Env(env), Napi::Object(env, exports), Napi::Object(env, module)); \
+    }                                                                                 \
+    catch (const Napi::Error&) {                                                      \
+      assert(false); /* Uncaught error in native module registration. */              \
+    }                                                                                 \
+  }                                                                                   \
+  NODE_MODULE_ABI(modname, __napi_ ## regfunc);
+
+
+////////////////////////////////////////////////////////////////////////////////
+// Env class
+////////////////////////////////////////////////////////////////////////////////
+
+inline Env::Env(napi_env env) : _env(env) {
+}
+
+inline Env::operator napi_env() const {
+  return _env;
+}
+inline Env Env::Current() {
+  napi_env env;
+  napi_status status = napi_get_current_env(&env);
+  if (status != napi_ok) env = nullptr; // Can't throw without an environment!
+  return Env(env);
+}
+
+inline Object Env::Global() const {
+  napi_value value;
```

Comment:
If I remember correctly, regarding V8, the primitives are not handled as normal pointers or such. They are special values of sorts:
```c++
Local<Primitive> Null(Isolate* isolate) {
  typedef internal::Object* S;
  typedef internal::Internals I;
  I::CheckInitialized(isolate);
  S* slot = I::GetRoot(isolate, I::kNullValueRootIndex);
  return Local<Primitive>(reinterpret_cast<Primitive*>(slot));
}
```
So as far as V8 is concerned, there should be no problem with caching primitives. I do not know how other engines handle primtives.